website/pages/docs/secrets/databases/custom.mdx

```diff
@@ -12,81 +12,166 @@ description: |-
 allows databases that require dynamically linked libraries to be used as plugins
 while keeping Vault itself statically linked.
 
-~> **Advanced topic!** Plugin development is a highly advanced topic in Vault,
-and is not required knowledge for day-to-day usage. If you don't plan on writing
-any plugins, we recommend not reading this section of the documentation.
-
 Please read the [Plugins internals](/docs/internals/plugins) docs for more
 information about the plugin system before getting started building your
 Database plugin.
 
 ## Plugin Interface
 
-All plugins for the database secrets engine must implement the same simple interface.
+All plugins for the database secrets engine must implement the same interface. This interface
+is found in `sdk/database/dbplugin/v5/database.go`
 
 ```go
 type Database interface {
+	// Initialize the database plugin. This is the equivalent of a constructor for the
+	// database object itself.
+	Initialize(ctx context.Context, req InitializeRequest) (InitializeResponse, error)
+
+	// NewUser creates a new user within the database. This user is temporary in that it
+	// will exist until the TTL expires.
+	NewUser(ctx context.Context, req NewUserRequest) (NewUserResponse, error)
+
+	// UpdateUser updates an existing user within the database.
+	UpdateUser(ctx context.Context, req UpdateUserRequest) (UpdateUserResponse, error)
+
+	// DeleteUser from the database. This should not error if the user didn't
+	// exist prior to this call.
+	DeleteUser(ctx context.Context, req DeleteUserRequest) (DeleteUserResponse, error)
+
+	// Type returns the Name for the particular database backend implementation.
+	// This type name is usually set as a constant within the database backend
+	// implementation, e.g. "mysql" for the MySQL database backend. This is used
+	// for things like metrics and logging. No behavior is switched on this.
 	Type() (string, error)
-	CreateUser(ctx context.Context, statements Statements, usernameConfig UsernameConfig, expiration time.Time) (username string, password string, err error)
-	RenewUser(ctx context.Context, statements Statements, username string, expiration time.Time) error
-	RevokeUser(ctx context.Context, statements Statements, username string) error
-	RotateRootCredentials(ctx context.Context, statements []string) (config map[string]interface{}, err error)
-	Init(ctx context.Context, config map[string]interface{}, verifyConnection bool) (saveConfig map[string]interface{}, err error)
+
+	// Close attempts to close the underlying database connection that was
+	// established by the backend.
 	Close() error
 }
 ```
 
-You'll notice the first parameter to a number of those functions is a
-`Statements` struct. This struct is used to pass the Role's configured
-statements to the plugin on function call. The struct is defined as:
+Each of the request and response objects can also be found in `sdk/database/dbplugin/v5/database.go`.
```

Comment:
Is there existing precedent for representing paths to code this way? 2 considerations:

1. Unrooted paths are sometimes confusing
2. This path makes more sense in the context of reading docs in the repo than the website. For someone reading the docs on the website, this path may be doubly confusing because it's both unrooted and it is in no way relative to the website.

OTOH I can see why you've not used absolute paths, because that would require using a branch or blob name, which would either link rot or quickly become out of date. Overall I think what you've done is probably my preferred option, but is there anything we can do to help clarify? Previously I've seen people use e.g. `vault.git/sdk/...`, which has its own flaws but another option.

Reply:
I don't think there is precedence for this.

1. Not sure what we can do for this. This page is explicitly for people who want to write a custom plugin and need to have knowledge about writing Go.
2. I disagree. The path is to a `.go` file and is in reference to writing a new plugin. I feel the context makes that clearer.

Yea, I agree there are alternatives to this but none of them are particularly good. Since the `sdk` package isn't going anywhere and we try to remain backwards compatible in that package, I feel pretty comfortable with this.